keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1235 @@
+perspective, the first set of scenarios would be for containers that can only
+override the restartPolicy to “higher” value. From Never to OnFailure or Always,
+from OnFailure to Always. An override from Always to Never will not be
+allowed.This will eliminate situations when kubelet restarts and cannot decide
+whether the container overridden value to Never will need to be restarted or not
+for the Pod with the restart policy Always. It is also likely that we will
+require at least one container to have a restartPolicy matching the Pod
+restartPolicy. We can re-evaluate this in future if scenarios will arise.
+
+Note, by implementing these scenarios, the meaning of restartPolicy:Always for
+containers on a Pod with the restartPolicy:Never, will not exactly be the
+meaning of the restartPolicy:Always for the Pod with the restartPolicy:Always.
+For the first case, containers will not block the Pod termination. For the
+second, they will. In other words, the container with restartPolicy:Always will
+effectively become a sidecar container similar to Init container with the same
+field value.
+
+Additional considerations can be found in the older document that was a previous
+attempt to implement the sidecar containers:
+https://docs.google.com/document/d/1gX_SOZXCNMcIe9d8CkekiT0GU9htH_my8OB7HSm3fM8/edit#heading=h.71o8fcesvgba.
+
+### Naming
+
+#### Collection name
+
+For this KEP it is important to have sidecar containers be defined among other
+Init containers to be able to express the initialization order of containers.
+The name `initContainers` is not a good fit for sidecar containers as they
+typically do more than initialization. The better name would be “infrastructure”
+containers. The current idea is to implement sidecars as a part of init
+containers and if this introduces too much trouble, the new collection name may
+replace the old collection name in future. 
+
+#### Reuse of restartPolicy field and enum
+
+The per-container restart policy was a long standing request from the community.
+Implementing per-container restart policy introduces a set of challenging
+problems for the pod lifecycle. For example, the state keeping for containers
+which has already run to completion. Introducing sidecar containers is another
+scenario where this field can be semantically used.
+
+Introducing this field on containers opens up opportunities to implement those
+long-standing requests from the community.
+
+#### Use “Always” vs. New enum value
+
+The semantic of an “Always” enum value is very close to what sidecars require.
+This is why reusing Always to represent sidecars makes a lot of sense for Init
+containers.
+
+Why “Always” that may not look similar to Always for other containers makes
+sense:
+
+Pros:
+- Allows the same enum values to be used for the `restartPolicy` of both pods
+and containers, but semantics that are mostly the same. Cons:
+- There are slight differences between the semantics of `Always` for containers
```

Comment:
Since these are two different fields (container restartPolicy and pod restartPolicy) with very similar semantic meaning (impacting behavior of containers), AND pod restartPolicy influences default container restartPolicy, I don't think this is a huge issue, and agree with your last comment.